Table by scanide

Scandie opened the pull request.

Scandie committed:
add first version of table view

Scandie committed:
add first working version of table functionality

*add Table and TableRow models
*add views for tables
*add templates for tables representation

wolendranh commented on rate_counter/rate/urls.py:
```diff
@@ -1,2 +1,27 @@
 from django.conf.urls import url
 
+from .views import (
+    create_table,
```
here you can save place with
`import .views`
and then
`view.create_table`

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,161 @@
+    """
+    creation of new table
+    """
+    form = TableForm(request.POST or None)
```
Where will we use this table? I mean once it is saved do we have some cases when it will be needed?

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,161 @@
+    """
+    table_obj = get_object_or_404(Table, id=id)
+    form = TableForm(request.POST or None, request.FILES or None, instance=table_obj)
+    if form.is_valid():
```
If not valid will it show errors on Form?

Scandie replied:
@wolendranh  It will show a message that unfilled input field is required.

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,161 @@
+    if form.is_valid():
+        instance = form.save(commit=False)
+        instance.save()
+        return HttpResponseRedirect("/rate/detail/%s/" % instance.id)
```
If you will use 'named' urls it will make this more cleaner. [Example](https://docs.djangoproject.com/en/dev/ref/urlresolvers/#django.urls.reverse) with reversing

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,161 @@
+    table_obj = get_object_or_404(Table, id=id)
+    table_rows = TableRow.objects.filter(table=table_obj)
+    coef_sum, point_sum, result = 0, 0, 0
+    if table_rows:
```
I suggest to move counting of rate to separate function. This will give you a possibility to cover it with unit test, For example in some modules like 'utils'

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,161 @@
+    for row in rows:
+        row.name = request.POST.get('name{}'.format(row.id))
+        row.coefficient = request.POST.get('kof{}'.format(row.id))
+        row.point = request.POST.get('bal{}'.format(row.id))
```
'bal' - seems like ukrainian word

wolendranh commented on rate_counter/templates/table_detail.html:
```diff
@@ -0,0 +1,83 @@
+        <table>
+            {% if rows %}
+                <tr>
+                    <td height="32px"><strong>Subject name</strong></td>
```
this is called 'inline' styles, when you write css inside HTML. Bad practice. I suggest to move this into separate file with styles.

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,161 @@
+    if form.is_valid():
+        instance = form.save(commit=False)
+        instance.save()
+        return HttpResponseRedirect("/rate/detail/%s/" % instance.id)
```
Better to use 'format' method. % is a C-style formatting, I think it was removed in Python 3

Scandie committed:
Add post-review changes

*add static directory for staticfiles
*add rate/utils.py for calculating functions
*

Scandie committed:
add new design to table page

wolendranh commented on rate_counter/rate/urls.py:
```diff
@@ -1,2 +1,19 @@
 from django.conf.urls import url
 
+import views
```
Nice. From my point of view now it is much more cleaner

wolendranh commented on rate_counter/rate/utils.py:
```diff
@@ -0,0 +1,19 @@
+def get_color(result):
```
@Scandie what do you think - does it make sense to write unit tests for your utils in this ticket or create separate one?

Scandie replied:
@wolendranh I think it is better to create a separate ticket for testing.

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,159 @@
+
+
+def delete_table(request, id=None):
+    """
```
@Scandie now it is totally ok to use function based views, because they are very small. In a future we might start use [Django class based views](https://docs.djangoproject.com/en/1.9/ref/class-based-views/generic-editing/#django.views.generic.edit.DeleteView). No need to do it now, just keep in mind.

wolendranh commented on rate_counter/rate/views.py:
```diff
@@ -1,3 +1,159 @@
+    """
+    table_obj = Table.objects.get(id=id)
+    TableRow.objects.create(table=table_obj, name='', coefficient=1, point=0)
+    return HttpResponseRedirect(reverse('rate:table_detail', args=[str(table_obj.id)]))
```
@Scandie Does it fails without casting to str here?

Scandie replied:
@wolendranh No, it is working well. Already removed it.

wolendranh commented on rate_counter/templates/table_detail.html:
```diff
@@ -0,0 +1,91 @@
+                        {% for row in rows %}
+                        <tr>
+                            <td><input size="120" class="form-control input-md" onchange="submit" value="{{row.name}}" name="name{{row.id}}" type="text"  maxlength="100"></td>
+                            <td><input class="form-control input-md" id="row_num_input" onchange="submit" value="{{row.coefficient}}" name="coef{{row.id}}" type="number" step="0.5" min="0" max="10"></td>
```
Here you are using same id selector in two same HTML elements. Id is for something identical :) here it is better to use  class. Also from my experience I am trying to avoid using id a lot.

Scandie replied:
@wolendranh Ok, I will keep it in mind.

wolendranh commented on rate_counter/templates/base.html:
```diff
@@ -0,0 +1,56 @@
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap-theme.min.css" integrity="sha384-fLW2N01lMqjakBkx3l/M9EahuwpSfeNvV63J5ezn3uZzapT0u7EYsXMjQV+0En5r" crossorigin="anonymous">
+
+    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>
+    <script src="//netdna.bootstrapcdn.com/bootstrap/3.1.1/js/bootstrap.min.js"></script>
```
Do we use bootstrap js in current state of project? It is ok to use it, but I am not sure if it is actually used somewhere in this project on this stage. Just be careful about js dependency - they grow like a cancer :)

Scandie replied:
@wolendranh It is used for navbar drop-down menu.

Scandie committed:
add some minor changes

*replace id "row_num_input" by class
*remove redundant str convertig

wolendranh commented:
Cool.
